module/module_test.go

```diff
@@ -723,6 +724,82 @@ func TestAttributeConversion(t *testing.T) {
 	})
 }
 
+// setupLocalModule sets up a module without a parent connection.
+func setupLocalModule(t *testing.T, ctx context.Context, logger logging.Logger) *module.Module {
+	t.Helper()
+
+	// Use 'foo.sock' for arbitrary module to test AddModelFromRegistry.
+	m, err := module.NewModule(ctx, filepath.Join(t.TempDir(), "foo.sock"), logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Hit Ready as a way to close m.pcFailed, so that AddResource can proceed. Set NoModuleParentEnvVar so that parent connection
+	// will not be attempted.
+	test.That(t, os.Setenv(module.NoModuleParentEnvVar, "true"), test.ShouldBeNil)
+	_, err = m.Ready(ctx, &pb.ReadyRequest{})
+	test.That(t, err, test.ShouldBeNil)
+
+	t.Cleanup(func() {
+		m.Close(ctx)
+		test.That(t, os.Unsetenv(module.NoModuleParentEnvVar), test.ShouldBeNil)
+	})
+	return m
+}
+
+// TestModuleAddResource tests that modular resources gets closed on add if context is done before resource finished creation.
+func TestModuleAddResource(t *testing.T) {
+	ctx := context.Background()
+
+	m := setupLocalModule(t, ctx, logging.NewTestLogger(t))
+
+	var (
+		shouldCancel bool
+
+		constructCount int
+		closeCount     int
+	)
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
minor, should we defer cancellation to ensure it gets called?

```suggestion
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
```

Reply:
done